TestCollectorToIntermediate is flaky

**Describe the bug**
TestCollectorToIntermediate is flaky. It is not passing consistently. Example run: https://github.com/vmware/go-ipfix/runs/1539468529

**To Reproduce**
Run the test locally with count=50. 

**Expected behavior**
Test should pass consistently.

**Actual Behavior**
I1211 19:59:29.646039    4907 tcp.go:38] Start tcp collecting process on [::]:4739
I1211 19:59:29.749036    4907 tcp.go:73] Connection from 127.0.0.1:49328 has been closed.
--- FAIL: TestCollectorToIntermediate (0.21s)
    collector_intermediate_test.go:167: 
        	Error Trace:	collector_intermediate_test.go:167
        	Error:      	Not equal: 
        	            	expected: 0x190
        	            	actual  : 0x3e8
        	Test:       	TestCollectorToIntermediate
    collector_intermediate_test.go:177: 
        	Error Trace:	collector_intermediate_test.go:177
        	Error:      	Not equal: 
        	            	expected: 0x3e8
        	            	actual  : 0x190
        	Test:       	TestCollectorToIntermediate

**Versions**
ToT master branch

--- a/.gitignore
+++ b/.gitignore
@@ -1,6 +1,7 @@
 bin
 pkg/registry/build_registry/build_registry
 .coverage/
+.golangci-bin/
 
 .DS_Store
 

--- a/Makefile
+++ b/Makefile
@@ -20,9 +20,12 @@ test-unit: .coverage
 test-integration: .coverage
 	$(GO) test -race ./pkg/test/... -tags=integration -covermode=atomic -coverprofile=.coverage/coverage_integration.txt -coverpkg github.com/vmware/go-ipfix/pkg/collector,github.com/vmware/go-ipfix/pkg/exporter,github.com/vmware/go-ipfix/pkg/intermediate
 
-golangci:
-	@curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(GOPATH)/bin v1.21.0
-	$(GOPATH)/bin/golangci-lint run -c .golangci.yml
+.golangci-bin:
+	@echo "===> Installing Golangci-lint <==="
+	@curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $@ v1.32.1
+
+golangci:.golangci-bin
+	@GOOS=linux .golangci-bin/golangci-lint run -c .golangci.yml
 
 collector:
 	@mkdir -p $(BINDIR)

--- a/cmd/collector/collector.go
+++ b/cmd/collector/collector.go
@@ -47,15 +47,10 @@ var (
 
 func initLoggingToFile(fs *pflag.FlagSet) {
 	var err error
-	var logToStdErr bool
 
-	logToStdErr, err = fs.GetBool(logToStdErrFlag)
+	_, err = fs.GetBool(logToStdErrFlag)
 	if err != nil {
 		// Should not happen. Return for safety.
-		return
-	}
-	if logToStdErr {
-		// Logging to files is not enabled.
 		return
 	}
 }

--- a/pkg/collector/process_test.go
+++ b/pkg/collector/process_test.go
@@ -178,7 +178,7 @@ func init() {
 }
 
 func TestTCPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
-	address, err := net.ResolveTCPAddr("tcp", "0.0.0.0:4730")
+	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -196,12 +196,12 @@ func TestTCPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
 	}
 	go cp.Start()
 	// wait until collector is ready
-	waitForCollectorReady(t, address)
-
+	waitForCollectorReady(t, cp)
+	collectorAddr := cp.GetAddress()
 	go func() {
-		conn, err := net.Dial(address.Network(), address.String())
+		conn, err := net.Dial(collectorAddr.Network(), collectorAddr.String())
 		if err != nil {
-			t.Errorf("Cannot establish connection to %s", address.String())
+			t.Errorf("Cannot establish connection to %s", collectorAddr.String())
 		}
 		defer conn.Close()
 		conn.Write(validTemplatePacket)
@@ -213,7 +213,7 @@ func TestTCPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
 }
 
 func TestUDPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
-	address, err := net.ResolveUDPAddr("udp", "0.0.0.0:4731")
+	address, err := net.ResolveUDPAddr("udp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -231,10 +231,10 @@ func TestUDPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
 	}
 	go cp.Start()
 	// wait until collector is ready
-	waitForCollectorReady(t, address)
-
+	waitForCollectorReady(t, cp)
+	collectorAddr := cp.GetAddress()
 	go func() {
-		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
+		resolveAddr, err := net.ResolveUDPAddr(collectorAddr.Network(), collectorAddr.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
 		}
@@ -253,7 +253,7 @@ func TestUDPCollectingProcess_ReceiveTemplateRecord(t *testing.T) {
 }
 
 func TestTCPCollectingProcess_ReceiveDataRecord(t *testing.T) {
-	address, err := net.ResolveTCPAddr("tcp", "0.0.0.0:4732")
+	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -275,12 +275,12 @@ func TestTCPCollectingProcess_ReceiveDataRecord(t *testing.T) {
 	go cp.Start()
 
 	// wait until collector is ready
-	waitForCollectorReady(t, address)
-
+	waitForCollectorReady(t, cp)
+	collectorAddr := cp.GetAddress()
 	go func() {
-		conn, err := net.Dial(address.Network(), address.String())
+		conn, err := net.Dial(collectorAddr.Network(), collectorAddr.String())
 		if err != nil {
-			t.Errorf("Cannot establish connection to %s", address.String())
+			t.Errorf("Cannot establish connection to %s", collectorAddr.String())
 		}
 		defer conn.Close()
 		conn.Write(validDataPacket)
@@ -290,7 +290,7 @@ func TestTCPCollectingProcess_ReceiveDataRecord(t *testing.T) {
 }
 
 func TestUDPCollectingProcess_ReceiveDataRecord(t *testing.T) {
-	address, err := net.ResolveUDPAddr("udp", "0.0.0.0:4733")
+	address, err := net.ResolveUDPAddr("udp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -311,10 +311,10 @@ func TestUDPCollectingProcess_ReceiveDataRecord(t *testing.T) {
 
 	go cp.Start()
 	// wait until collector is ready
-	waitForCollectorReady(t, address)
-
+	waitForCollectorReady(t, cp)
+	collectorAddr := cp.GetAddress()
 	go func() {
-		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
+		resolveAddr, err := net.ResolveUDPAddr(collectorAddr.Network(), collectorAddr.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
 		}
@@ -330,7 +330,7 @@ func TestUDPCollectingProcess_ReceiveDataRecord(t *testing.T) {
 }
 
 func TestTCPCollectingProcess_ConcurrentClient(t *testing.T) {
-	address, err := net.ResolveTCPAddr("tcp", "0.0.0.0:4734")
+	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -345,28 +345,30 @@ func TestTCPCollectingProcess_ConcurrentClient(t *testing.T) {
 	cp, _ := InitCollectingProcess(input)
 	go func() {
 		// wait until collector is ready
-		waitForCollectorReady(t, address)
-		_, err := net.Dial(address.Network(), address.String())
+		waitForCollectorReady(t, cp)
+		collectorAddr := cp.GetAddress()
+		_, err := net.Dial(collectorAddr.Network(), collectorAddr.String())
 		if err != nil {
-			t.Errorf("Cannot establish connection to %s", address.String())
+			t.Errorf("Cannot establish connection to %s", collectorAddr.String())
 		}
 	}()
 	go func() {
 		// wait until collector is ready
-		waitForCollectorReady(t, address)
-		_, err := net.Dial(address.Network(), address.String())
+		waitForCollectorReady(t, cp)
+		collectorAddr := cp.GetAddress()
+		_, err := net.Dial(collectorAddr.Network(), collectorAddr.String())
 		if err != nil {
-			t.Errorf("Cannot establish connection to %s", address.String())
+			t.Errorf("Cannot establish connection to %s", collectorAddr.String())
 		}
 		time.Sleep(time.Millisecond)
-		assert.Equal(t, 4, cp.getClientCount(), "There should be 4 tcp clients.")
+		assert.Equal(t, 2, cp.getClientCount(), "There should be two tcp clients.")
 		cp.Stop()
 	}()
 	cp.Start()
 }
 
 func TestUDPCollectingProcess_ConcurrentClient(t *testing.T) {
-	address, err := net.ResolveUDPAddr("udp", "0.0.0.0:4735")
+	address, err := net.ResolveUDPAddr("udp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -381,9 +383,10 @@ func TestUDPCollectingProcess_ConcurrentClient(t *testing.T) {
 	cp, _ := InitCollectingProcess(input)
 	go cp.Start()
 	// wait until collector is ready
-	waitForCollectorReady(t, address)
+	waitForCollectorReady(t, cp)
+	collectorAddr := cp.GetAddress()
 	go func() {
-		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
+		resolveAddr, err := net.ResolveUDPAddr(collectorAddr.Network(), collectorAddr.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
 		}
@@ -395,7 +398,7 @@ func TestUDPCollectingProcess_ConcurrentClient(t *testing.T) {
 		conn.Write(validTemplatePacket)
 	}()
 	go func() {
-		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
+		resolveAddr, err := net.ResolveUDPAddr(collectorAddr.Network(), collectorAddr.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
 		}
@@ -406,8 +409,7 @@ func TestUDPCollectingProcess_ConcurrentClient(t *testing.T) {
 		defer conn.Close()
 		conn.Write(validTemplatePacket)
 		time.Sleep(time.Millisecond)
-		// function waitForCollectorReady may introduce more clients when testing connection
-		assert.GreaterOrEqual(t, 2, cp.getClientCount(), "There should be at least two tcp clients.")
+		assert.Equal(t, 2, cp.getClientCount(), "There should be two udp clients.")
 	}()
 	// there should be two messages received
 	<-cp.GetMsgChan()
@@ -419,7 +421,7 @@ func TestCollectingProcess_DecodeTemplateRecord(t *testing.T) {
 	cp := CollectingProcess{}
 	cp.templatesMap = make(map[uint32]map[uint16][]*entities.InfoElement)
 	cp.mutex = sync.RWMutex{}
-	address, err := net.ResolveTCPAddr("tcp", "0.0.0.0:4736")
+	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -460,7 +462,7 @@ func TestCollectingProcess_DecodeDataRecord(t *testing.T) {
 	cp := CollectingProcess{}
 	cp.templatesMap = make(map[uint32]map[uint16][]*entities.InfoElement)
 	cp.mutex = sync.RWMutex{}
-	address, err := net.ResolveTCPAddr("tcp", "0.0.0.0:4737")
+	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -493,7 +495,7 @@ func TestCollectingProcess_DecodeDataRecord(t *testing.T) {
 }
 
 func TestUDPCollectingProcess_TemplateExpire(t *testing.T) {
-	address, err := net.ResolveUDPAddr("udp", "0.0.0.0:4738")
+	address, err := net.ResolveUDPAddr("udp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -511,9 +513,10 @@ func TestUDPCollectingProcess_TemplateExpire(t *testing.T) {
 	}
 	go cp.Start()
 	// wait until collector is ready
-	waitForCollectorReady(t, address)
+	waitForCollectorReady(t, cp)
+	collectorAddr := cp.GetAddress()
 	go func() {
-		resolveAddr, err := net.ResolveUDPAddr(address.Network(), address.String())
+		resolveAddr, err := net.ResolveUDPAddr(collectorAddr.Network(), collectorAddr.String())
 		if err != nil {
 			t.Errorf("UDP Address cannot be resolved.")
 		}
@@ -539,7 +542,7 @@ func TestUDPCollectingProcess_TemplateExpire(t *testing.T) {
 }
 
 func TestTLSCollectingProcess(t *testing.T) {
-	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:4739")
+	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -558,7 +561,8 @@ func TestTLSCollectingProcess(t *testing.T) {
 	}
 	go cp.Start()
 	// wait until collector is ready
-	waitForCollectorReady(t, address)
+	waitForCollectorReady(t, cp)
+	collectorAddr := cp.GetAddress()
 	go func() {
 		roots := x509.NewCertPool()
 		ok := roots.AppendCertsFromPEM([]byte(fakeCACert))
@@ -574,7 +578,7 @@ func TestTLSCollectingProcess(t *testing.T) {
 			Certificates: []tls.Certificate{cert},
 		}
 
-		conn, err := tls.Dial("tcp", address.String(), config)
+		conn, err := tls.Dial("tcp", collectorAddr.String(), config)
 		if err != nil {
 			t.Error(err)
 			return
@@ -589,7 +593,7 @@ func TestTLSCollectingProcess(t *testing.T) {
 }
 
 func TestDTLSCollectingProcess(t *testing.T) {
-	address, err := net.ResolveUDPAddr("udp", "0.0.0.0:4740")
+	address, err := net.ResolveUDPAddr("udp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -607,7 +611,8 @@ func TestDTLSCollectingProcess(t *testing.T) {
 	}
 	go cp.Start()
 	// wait until collector is ready
-	waitForCollectorReady(t, address)
+	waitForCollectorReady(t, cp)
+	collectorAddr, _ := net.ResolveUDPAddr("udp", cp.GetAddress().String())
 	go func() {
 		roots := x509.NewCertPool()
 		ok := roots.AppendCertsFromPEM([]byte(fakeCert2))
@@ -617,7 +622,7 @@ func TestDTLSCollectingProcess(t *testing.T) {
 		config := &dtls.Config{RootCAs: roots,
 			ExtendedMasterSecret: dtls.RequireExtendedMasterSecret}
 
-		conn, err := dtls.Dial("udp", address, config)
+		conn, err := dtls.Dial("udp", collectorAddr, config)
 		if err != nil {
 			t.Error(err)
 			return
@@ -631,14 +636,16 @@ func TestDTLSCollectingProcess(t *testing.T) {
 	assert.NotNil(t, cp.templatesMap[1], "DTLS Collecting Process should receive and store the received template.")
 }
 
-func waitForCollectorReady(t *testing.T, address net.Addr) {
+func waitForCollectorReady(t *testing.T, cp *CollectingProcess) {
 	checkConn := func() (bool, error) {
-		if _, err := net.Dial(address.Network(), address.String()); err != nil {
+		if conn, err := net.Dial(cp.GetAddress().Network(), cp.GetAddress().String()); err != nil {
 			return false, err
+		} else {
+			defer conn.Close()
+			return true, nil
 		}
-		return true, nil
 	}
 	if err := wait.Poll(100*time.Millisecond, 500*time.Millisecond, checkConn); err != nil {
-		t.Errorf("Cannot establish connection to %s", address.String())
+		t.Errorf("Cannot establish connection to %s", cp.GetAddress().String())
 	}
 }

--- a/pkg/collector/tcp.go
+++ b/pkg/collector/tcp.go
@@ -7,7 +7,6 @@ import (
 	"fmt"
 	"io"
 	"net"
-	"sync"
 
 	"k8s.io/klog"
 )
@@ -38,7 +37,6 @@ func (cp *CollectingProcess) startTCPServer() {
 		klog.Infof("Start %s collecting process on %s", cp.address.Network(), cp.address.String())
 	}
 
-	var wg sync.WaitGroup
 	go func() {
 		defer listener.Close()
 		for {
@@ -47,18 +45,15 @@ func (cp *CollectingProcess) startTCPServer() {
 				klog.Errorf("Cannot start collecting process on %s: %v", cp.address.String(), err)
 				return
 			}
-			wg.Add(1)
-			go cp.handleTCPClient(conn, &wg)
+			go cp.handleTCPClient(conn)
 		}
 	}()
 	<-cp.stopChan
 	// close all connections
 	cp.closeAllClients()
-	wg.Wait()
 }
 
-func (cp *CollectingProcess) handleTCPClient(conn net.Conn, wg *sync.WaitGroup) {
-	defer wg.Done()
+func (cp *CollectingProcess) handleTCPClient(conn net.Conn) {
 	address := conn.RemoteAddr().String()
 	client := cp.createClient()
 	cp.addClient(address, client)
@@ -74,6 +69,7 @@ func (cp *CollectingProcess) handleTCPClient(conn net.Conn, wg *sync.WaitGroup) 
 				} else {
 					klog.Errorf("Error in collecting process: %v", err)
 				}
+				client.errChan <- true
 				break out
 			}
 			klog.V(2).Infof("Receiving %d bytes from %s", size, address)
@@ -81,6 +77,7 @@ func (cp *CollectingProcess) handleTCPClient(conn net.Conn, wg *sync.WaitGroup) 
 				length, err := getMessageLength(bytes.NewBuffer(buff))
 				if err != nil {
 					klog.Error(err)
+					client.errChan <- true
 					break out
 				}
 				if size < length {
@@ -92,6 +89,7 @@ func (cp *CollectingProcess) handleTCPClient(conn net.Conn, wg *sync.WaitGroup) 
 				message, err := cp.decodePacket(bytes.NewBuffer(buff[0:length]), address)
 				if err != nil {
 					klog.Error(err)
+					client.errChan <- true
 					break out
 				}
 				klog.V(4).Info(message)
@@ -111,6 +109,7 @@ func (cp *CollectingProcess) createServerConfig() (*tls.Config, error) {
 	if cp.caCert == nil {
 		return &tls.Config{
 			Certificates: []tls.Certificate{cert},
+			MinVersion:   tls.VersionTLS12,
 		}, nil
 	}
 	roots := x509.NewCertPool()
@@ -122,5 +121,6 @@ func (cp *CollectingProcess) createServerConfig() (*tls.Config, error) {
 		Certificates: []tls.Certificate{cert},
 		ClientAuth:   tls.RequireAndVerifyClientCert,
 		ClientCAs:    roots,
+		MinVersion:   tls.VersionTLS12,
 	}, nil
 }

--- a/pkg/collector/udp.go
+++ b/pkg/collector/udp.go
@@ -93,7 +93,6 @@ func (cp *CollectingProcess) startUDPServer() {
 		}
 		cp.updateAddress(conn.LocalAddr())
 		klog.Infof("Start %s collecting process on %s", cp.address.Network(), cp.address.String())
-		var wg sync.WaitGroup
 		defer conn.Close()
 		go func() {
 			for {

--- a/pkg/exporter/process.go
+++ b/pkg/exporter/process.go
@@ -340,7 +340,8 @@ func createClientConfig(caCert, clientCert, clientKey []byte) (*tls.Config, erro
 	}
 	if clientCert == nil {
 		return &tls.Config{
-			RootCAs: roots,
+			RootCAs:    roots,
+			MinVersion: tls.VersionTLS12,
 		}, nil
 	}
 	cert, err := tls.X509KeyPair(clientCert, clientKey)
@@ -350,5 +351,6 @@ func createClientConfig(caCert, clientCert, clientKey []byte) (*tls.Config, erro
 	return &tls.Config{
 		Certificates: []tls.Certificate{cert},
 		RootCAs:      roots,
+		MinVersion:   tls.VersionTLS12,
 	}, nil
 }

--- a/pkg/exporter/process_test.go
+++ b/pkg/exporter/process_test.go
@@ -471,7 +471,7 @@ func TestExportingProcess_SendingDataRecordToLocalUDPServer(t *testing.T) {
 
 func TestExportingProcessWithTLS(t *testing.T) {
 	// Create local server for testing
-	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:4739")
+	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 	if err != nil {
 		t.Fatalf("Got error when resolving tcp address: %v", err)
 	}
@@ -508,7 +508,7 @@ func TestExportingProcessWithTLS(t *testing.T) {
 
 	// Create exporter using local server info
 	input := ExporterInput{
-		CollectorAddr:       address,
+		CollectorAddr:       listener.Addr(),
 		ObservationDomainID: 1,
 		TempRefTimeout:      0,
 		IsEncrypted:         true,
@@ -516,9 +516,9 @@ func TestExportingProcessWithTLS(t *testing.T) {
 	}
 	exporter, err := InitExportingProcess(input)
 	if err != nil {
-		t.Fatalf("Got error when connecting to local tls server %s: %v", address, err)
+		t.Fatalf("Got error when connecting to local tls server %s: %v", listener.Addr(), err)
 	}
-	t.Logf("Created exporter connecting to local tls server with address: %s", address)
+	t.Logf("Created exporter connecting to local tls server with address: %s", listener.Addr())
 
 	// Create template record with two fields
 	templateID := exporter.NewTemplateID()
@@ -550,7 +550,7 @@ func TestExportingProcessWithTLS(t *testing.T) {
 
 func TestExportingProcessWithDTLS(t *testing.T) {
 	// Create local server for testing
-	address, err := net.ResolveUDPAddr("udp", "0.0.0.0:4831")
+	address, err := net.ResolveUDPAddr("udp", "127.0.0.1:0")
 	if err != nil {
 		t.Fatalf("Got error when resolving udp address: %v", err)
 	}
@@ -565,7 +565,7 @@ func TestExportingProcessWithDTLS(t *testing.T) {
 	}
 	listener, err := dtls.Listen("udp", address, config)
 	if err != nil {
-		t.Errorf("Cannot start dtls collecting process on 0.0.0.0:4739: %v", err)
+		t.Errorf("Cannot start dtls collecting process on %s: %v", listener.Addr().String(), err)
 		return
 	}
 
@@ -590,7 +590,7 @@ func TestExportingProcessWithDTLS(t *testing.T) {
 
 	// Create exporter using local server info
 	input := ExporterInput{
-		CollectorAddr:       address,
+		CollectorAddr:       listener.Addr(),
 		ObservationDomainID: 1,
 		TempRefTimeout:      0,
 		IsEncrypted:         true,
@@ -598,9 +598,9 @@ func TestExportingProcessWithDTLS(t *testing.T) {
 	}
 	exporter, err := InitExportingProcess(input)
 	if err != nil {
-		t.Fatalf("Got error when connecting to local dtls server %s: %v", address, err)
+		t.Fatalf("Got error when connecting to local dtls server %s: %v", listener.Addr().String(), err)
 	}
-	t.Logf("Created exporter connecting to local dtls server with address: %s", address)
+	t.Logf("Created exporter connecting to local dtls server with address: %s", listener.Addr().String())
 
 	// Create template record with two fields
 	templateID := exporter.NewTemplateID()

--- a/pkg/intermediate/aggregate.go
+++ b/pkg/intermediate/aggregate.go
@@ -124,8 +124,8 @@ func (a *AggregationProcess) AggregateMsgByFlowKey(message *entities.Message) er
 
 // ForAllRecordsDo takes in callback function to process the operations to flowkey->records pairs in the map
 func (a *AggregationProcess) ForAllRecordsDo(callback FlowKeyRecordMapCallBack) error {
-	a.mutex.RLock()
-	defer a.mutex.RUnlock()
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
 	for k, v := range a.flowKeyRecordMap {
 		err := callback(k, v)
 		if err != nil {
@@ -301,7 +301,9 @@ func (a *AggregationProcess) aggregateRecords(incomingRecord, existingRecord ent
 			existingIeWithValue, _ := existingRecord.GetInfoElementWithValue(element)
 			// Update the corresponding element in existing record.
 			if !isDelta {
-				existingIeWithValue.Value = ieWithValue.Value
+				if existingIeWithValue.Value.(uint64) < ieWithValue.Value.(uint64) {
+					existingIeWithValue.Value = ieWithValue.Value
+				}
 			} else {
 				// We are simply adding the delta stats now. We expect delta stats to be
 				// reset after sending the record from flowKeyMap in aggregation process.

--- a/pkg/registry/registry_test.go
+++ b/pkg/registry/registry_test.go
@@ -22,10 +22,11 @@ import (
 	"github.com/vmware/go-ipfix/pkg/entities"
 )
 
-func TestLoadRegistry(t *testing.T) {
-	assert.Equal(t, 0, len(globalRegistryByID))
-	assert.Equal(t, 0, len(globalRegistryByName))
+func init() {
 	LoadRegistry()
+}
+
+func TestLoadRegistry(t *testing.T) {
 	assert.NotEmpty(t, globalRegistryByName[IANAEnterpriseID])
 	assert.NotEmpty(t, globalRegistryByName[AntreaEnterpriseID])
 	assert.NotEmpty(t, globalRegistryByName[IANAReversedEnterpriseID])
@@ -35,7 +36,6 @@ func TestLoadRegistry(t *testing.T) {
 }
 
 func TestGetInfoElement(t *testing.T) {
-	LoadRegistry()
 	var expectedIE *entities.InfoElement
 	ie, error := GetInfoElement("ingressInterfaceTest", IANAEnterpriseID)
 	assert.Equal(t, expectedIE, ie, "GetIANAInfoElement did not return correct value.")
@@ -47,7 +47,6 @@ func TestGetInfoElement(t *testing.T) {
 }
 
 func TestGetIANAReverseIE(t *testing.T) {
-	LoadRegistry()
 	// InfoElement does not exist in the registry
 	_, error := getIANAReverseInfoElement("sourcePodName")
 	assert.NotEqual(t, nil, error, "GetIANAReverseIE should return error when ie does not exist.")
@@ -61,7 +60,6 @@ func TestGetIANAReverseIE(t *testing.T) {
 }
 
 func TestGetInfoElementFromID(t *testing.T) {
-	LoadRegistry()
 	// InfoElement does not exist
 	_, err := GetInfoElementFromID(1, 1)
 	assert.NotEqual(t, nil, err, "TestGetInfoElementFromID should return error when ie does not exist.")

--- a/pkg/test/collector_intermediate_test.go
+++ b/pkg/test/collector_intermediate_test.go
@@ -30,8 +30,8 @@ import (
 )
 
 var templatePacket = []byte{0, 10, 0, 104, 95, 208, 69, 75, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 0, 88, 1, 0, 0, 14, 0, 8, 0, 4, 0, 12, 0, 4, 0, 7, 0, 2, 0, 11, 0, 2, 0, 4, 0, 1, 0, 151, 0, 4, 0, 86, 0, 8, 0, 2, 0, 8, 128, 101, 255, 255, 0, 0, 220, 186, 128, 103, 255, 255, 0, 0, 220, 186, 128, 106, 0, 4, 0, 0, 220, 186, 128, 108, 0, 2, 0, 0, 220, 186, 128, 86, 0, 8, 0, 0, 114, 121, 128, 2, 0, 8, 0, 0, 114, 121}
-var dataPacket1 = []byte{0, 10, 0, 81, 95, 208, 77, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 65, 10, 0, 0, 1, 10, 0, 0, 2, 4, 210, 22, 46, 6, 74, 249, 240, 112, 0, 0, 0, 0, 0, 0, 3, 232, 0, 0, 0, 0, 0, 0, 1, 244, 0, 4, 112, 111, 100, 50, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 144, 0, 0, 0, 0, 0, 0, 0, 200}
-var dataPacket2 = []byte{0, 10, 0, 81, 95, 208, 77, 154, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 65, 10, 0, 0, 1, 10, 0, 0, 2, 4, 210, 22, 46, 6, 74, 249, 244, 88, 0, 0, 0, 0, 0, 0, 1, 144, 0, 0, 0, 0, 0, 0, 0, 200, 4, 112, 111, 100, 49, 0, 10, 0, 0, 3, 18, 131, 0, 0, 0, 0, 0, 0, 3, 232, 0, 0, 0, 0, 0, 0, 1, 244}
+var dataPacket1 = []byte{0, 10, 0, 81, 95, 218, 99, 200, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 65, 10, 0, 0, 1, 10, 0, 0, 2, 4, 210, 22, 46, 6, 74, 249, 240, 112, 0, 0, 0, 0, 0, 0, 3, 232, 0, 0, 0, 0, 0, 0, 1, 244, 0, 4, 112, 111, 100, 50, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 144, 0, 0, 0, 0, 0, 0, 0, 200}
+var dataPacket2 = []byte{0, 10, 0, 81, 95, 218, 100, 162, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 65, 10, 0, 0, 1, 10, 0, 0, 2, 4, 210, 22, 46, 6, 74, 249, 248, 64, 0, 0, 0, 0, 0, 0, 3, 32, 0, 0, 0, 0, 0, 0, 1, 244, 4, 112, 111, 100, 49, 0, 10, 0, 0, 3, 18, 131, 0, 0, 0, 0, 0, 0, 1, 44, 0, 0, 0, 0, 0, 0, 0, 150}
 
 /*
 dataPacket1:
@@ -55,15 +55,15 @@ dataPacket 2:
 	"sourceTransportPort": 1234
 	"destinationTransportPort": 5678
 	"protocolIdentifier": 6
-	"flowEndSeconds": 1257895000
-	"packetTotalCount": 400
-	"packetDeltaCount": 200
+	"flowEndSeconds": 1257896000
+	"packetTotalCount": 800
+	"packetDeltaCount": 500
 	"sourcePodName": "pod1"
 	"destinationPodName": ""
 	"destinationClusterIPv4": 10.0.0.3
 	"destinationServicePort": 4739
-	"reversePacketTotalCount": 1000
-	"reversePacketDeltaCount": 500
+	"reversePacketTotalCount": 300
+	"reversePacketDeltaCount": 150
 */
 
 var (
@@ -100,18 +100,22 @@ var (
 		"reversePacketTotalCountFromDestinationNode",
 		"reversePacketDeltaCountFromDestinationNode",
 	}
+	aggregationWorkerNum = 2
 )
 
-func TestCollectorToIntermediate(t *testing.T) {
+func init() {
+	// Load the global registry
 	registry.LoadRegistry()
+}
 
+func TestCollectorToIntermediate(t *testing.T) {
 	aggregatedFields := &intermediate.AggregationElements{
 		NonStatsElements:                   nonStatsElementList,
 		StatsElements:                      statsElementList,
 		AggregatedSourceStatsElements:      antreaSourceStatsElementList,
 		AggregatedDestinationStatsElements: antreaDestinationStatsElementList,
 	}
-	address, err := net.ResolveUDPAddr("udp", "0.0.0.0:4739")
+	address, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
 	if err != nil {
 		t.Error(err)
 	}
@@ -128,7 +132,7 @@ func TestCollectorToIntermediate(t *testing.T) {
 
 	apInput := intermediate.AggregationInput{
 		MessageChan:       cp.GetMsgChan(),
-		WorkerNum:         2,
+		WorkerNum:         aggregationWorkerNum,
 		CorrelateFields:   correlatefields,
 		AggregateElements: aggregatedFields,
 	}
@@ -136,9 +140,10 @@ func TestCollectorToIntermediate(t *testing.T) {
 	go cp.Start()
 	waitForCollectorReady(t, cp)
 	go func() {
-		conn, err := net.DialUDP("udp", nil, address)
+		collectorAddr, _ := net.ResolveTCPAddr("tcp", cp.GetAddress().String())
+		conn, err := net.DialTCP("tcp", nil, collectorAddr)
 		if err != nil {
-			t.Errorf("UDP Collecting Process does not start correctly.")
+			t.Errorf("TCP Collecting Process does not start correctly.")
 		}
 		defer conn.Close()
 		conn.Write(templatePacket)
@@ -162,25 +167,35 @@ func TestCollectorToIntermediate(t *testing.T) {
 		case "destinationPodName":
 			assert.Equal(t, "pod2", element.Value)
 		case "flowEndSeconds":
-			assert.Equal(t, uint32(1257895000), element.Value)
+			assert.Equal(t, uint32(1257896000), element.Value)
 		case "packetTotalCount":
-			assert.Equal(t, uint64(400), element.Value)
+			assert.Equal(t, uint64(1000), element.Value)
 		case "packetDeltaCount":
-			assert.Equal(t, uint64(700), element.Value)
+			assert.Equal(t, uint64(1000), element.Value)
 		case "destinationClusterIPv4":
 			assert.Equal(t, net.IP{10, 0, 0, 3}, element.Value)
 		case "destinationServicePort":
 			assert.Equal(t, uint16(4739), element.Value)
 		case "reversePacketDeltaCount":
-			assert.Equal(t, uint64(700), element.Value)
+			assert.Equal(t, uint64(350), element.Value)
 		case "reversePacketTotalCount":
-			assert.Equal(t, uint64(1000), element.Value)
-		case "packetTotalCountFromSourceNode":
 			assert.Equal(t, uint64(400), element.Value)
+		case "packetTotalCountFromSourceNode":
+			assert.Equal(t, uint64(800), element.Value)
 		case "packetDeltaCountFromSourceNode":
-			assert.Equal(t, uint64(200), element.Value)
+			assert.Equal(t, uint64(500), element.Value)
 		case "packetTotalCountFromDestinationNode":
 			assert.Equal(t, uint64(1000), element.Value)
+		case "packetDeltaCountFromDestinationNode":
+			assert.Equal(t, uint64(500), element.Value)
+		case "reversePacketTotalCountFromSourceNode":
+			assert.Equal(t, uint64(300), element.Value)
+		case "reversePacketDeltaCountFromSourceNode":
+			assert.Equal(t, uint64(150), element.Value)
+		case "reversePacketTotalCountFromDestinationNode":
+			assert.Equal(t, uint64(400), element.Value)
+		case "reversePacketDeltaCountFromDestinationNode":
+			assert.Equal(t, uint64(200), element.Value)
 		}
 	}
 

